quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/KeycloakMain.java

```diff
@@ -159,24 +159,20 @@ private void createAdminUser() {
         }
 
         KeycloakSessionFactory sessionFactory = KeycloakApplication.getSessionFactory();
-        KeycloakSession session = sessionFactory.create();
-        KeycloakTransactionManager transaction = session.getTransactionManager();
 
-        try {
+        try (KeycloakSession session = sessionFactory.create()) {
+            KeycloakTransactionManager transaction = session.getTransactionManager();
             transaction.begin();
-
-            new ApplianceBootstrap(session).createMasterRealmUser(adminUserName, adminPassword);
-            ServicesLogger.LOGGER.addUserSuccess(adminUserName, Config.getAdminRealm());
-
-            transaction.commit();
-        } catch (IllegalStateException e) {
-            session.getTransactionManager().rollback();
-            ServicesLogger.LOGGER.addUserFailedUserExists(adminUserName, Config.getAdminRealm());
-        } catch (Throwable t) {
-            session.getTransactionManager().rollback();
-            ServicesLogger.LOGGER.addUserFailed(t, adminUserName, Config.getAdminRealm());
-        } finally {
-            session.close();
+            try {
+                new ApplianceBootstrap(session).createMasterRealmUser(adminUserName, adminPassword);
+                ServicesLogger.LOGGER.addUserSuccess(adminUserName, Config.getAdminRealm());
+            } catch (IllegalStateException e) {
+                session.getTransactionManager().setRollbackOnly();
+                ServicesLogger.LOGGER.addUserFailedUserExists(adminUserName, Config.getAdminRealm());
```

Comment:
Since the transaction is not closed until after this is executed I don't see how it will get the exception if the user already exists.

Reply:
Good catch. Changed to `runJobInTransaction`.